crates/jstzd/src/protocol/mod.rs

```diff
@@ -0,0 +1,166 @@
+use rust_embed::Embed;
+use serde_json::Value;
+use std::fmt::Display;
+use std::path::{Path, PathBuf};
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum ProtocolConstants {
+    Sandbox,
+    #[cfg(test)]
+    TestConstants,
+}
+
+impl Default for ProtocolConstants {
+    fn default() -> Self {
+        Self::Sandbox
+    }
+}
+
+impl Display for ProtocolConstants {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(match self {
+            Self::Sandbox => "sandbox",
+            #[cfg(test)]
+            Self::TestConstants => "test-test",
+        })
+    }
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum Protocol {
+    Alpha,
+    #[cfg(test)]
+    TestVersion,
+}
+
+impl Default for Protocol {
+    fn default() -> Self {
+        Self::Alpha
+    }
+}
+
+impl Protocol {
+    fn hash(self) -> &'static str {
+        match self {
+            Protocol::Alpha => "ProtoALphaALphaALphaALphaALphaALphaALphaALphaDdp3zK",
+            #[cfg(test)]
+            Protocol::TestVersion => "test-hash",
+        }
+    }
+
+    fn parameter_file(self, constants: ProtocolConstants) -> PathBuf {
+        Path::new(&constants.to_string()).join(self.hash())
+    }
+}
+
+#[cfg(not(test))]
+#[derive(Embed)]
+#[folder = "$CARGO_MANIFEST_DIR/resources/protocol_parameters/"]
+pub struct ProtocolParameterFile;
+
+#[cfg(test)]
+#[derive(Embed)]
+#[folder = "$CARGO_MANIFEST_DIR/resources/test/protocol_parameters/"]
+pub struct ProtocolParameterFile;
+
+#[derive(Default)]
+pub struct ProtocolParameterBuilder {
+    protocol: Protocol,
+    constants: ProtocolConstants,
+    path: Option<PathBuf>,
+}
+
+impl ProtocolParameterBuilder {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn set_protocol(&mut self, protocol: Protocol) -> &mut Self {
+        self.protocol = protocol;
+        self
+    }
+
+    pub fn set_constants(&mut self, constants: ProtocolConstants) -> &mut Self {
+        self.constants = constants;
+        self
+    }
+
+    pub fn set_path(&mut self, path: &str) -> &mut Self {
+        self.path = Some(PathBuf::from(path));
+        self
+    }
+
+    pub async fn build(self) -> anyhow::Result<PathBuf> {
```

Comment:
I think maybe there's a misunderstanding what this function does. If the caller supplies a parameters file then it should use that one, otherwise we will default to our sandbox params file. Then, any values set on the builder should take precedence over the values set in the parameters file.

Reply:
Agreed, the `set_protocol` and `set_constants `would mainly be used for our internal framework since we won't be passing in specific params files but will be using the base params files provided. Perhaps it should be an error if set_protocol differs from the supplied parameters file?

Reply:
But there's no way of knowing which protocol version the provided parameter file is for. octez node does reject parameters that are invalid for a given protocol version, but if all of them are valid, there is no way of determining if the intended protocol version is different from the one updated by `set_protocol`.

Reply:
Ah yes, my bad, thought the protocol was in the parameters file. 

No-op is fine.